File: s3_20_searchingForStrings.java
import java.util.*;
import java.io.*;

public class s3_20_searchingForStrings {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int count = 0;
    static char[] letters;
    static boolean[] used;
    static Set<String> combos;
    static int nLen;

    public static void main(String[] args) throws IOException {
        String n = readLine();
        String h = readLine();
        nLen = n.length();

        if(nLen > h.length()) {
            System.out.println(0);
            return;
        }

        letters = n.toCharArray();
        used = new boolean[nLen];
        combos = new HashSet<>();

        findCombos("");

        for(String s : combos) {
            if(h.contains(s)) {
                count++;
            }
        }
        System.out.println(count);

    }

    static void findCombos(String curr) {
        if(curr.length() == nLen) {
            combos.add(curr);
            return;
        }
        for(int i = 0; i < nLen; i++) {
            if(!used[i]) {
                used[i] = true;
                findCombos(curr + letters[i]);
                used[i] = false;
            }
        }

    }


    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}

File: s3_20_searchingForStrings.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

static int count = 0;
static string n;
static vector<bool> used;
static unordered_set<string> combos;
static int nLen;

void findCombos(string curr) {
    if(curr.length() == nLen) {
        combos.insert(curr);
        return;
    }
    for(int i = 0; i < nLen; i++) {
        if(!used[i]) {
            used[i] = true;
            findCombos(curr + n[i]);
            used[i] = false;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    getline(cin, n);
    string h;
    getline(cin, h);
    nLen = n.length();
    
    if(nLen > h.length()) {
        cout << 0 << endl;
        return 0;
    }
    
    used.resize(nLen);
    findCombos("");
    
    for(const string& s : combos) {
        if(h.find(s) != string::npos) {
            count++;
        }
    }
    cout << count << endl;
    
    return 0;
}
